apps/heureka/src/components/services/ComponentInstancesList.jsx

```diff
@@ -0,0 +1,123 @@
+  const queryClientFnReady = useGlobalsQueryClientFnReady()
+  const [currentPage, setCurrentPage] = useState(1)
+
+  const { data, isLoading } = useQuery({
+    queryKey: [
+      "ComponentInstancesOfService",
+      {
+        ...queryOptions,
+        filter: { serviceName: [serviceName] },
+      },
+    ],
+    enabled: !!queryClientFnReady && !!serviceName,
+  })
+
+  const items = useMemo(() => {
+    return data?.ComponentInstances?.edges || []
+  }, [data])
+
+  const pageInfo = useMemo(() => {
+    return data?.ComponentInstances?.pageInfo || {}
+  }, [data])
+
+  const totalPages = useMemo(() => {
+    if (!pageInfo?.pages) return 0
+    return pageInfo.pages.length
+  }, [pageInfo])
+
+  const onPaginationChanged = (newPage) => {
+    setCurrentPage(newPage)
+    if (!pageInfo?.pages) return
+    const pages = pageInfo.pages
+    const currentPageIndex = pages?.findIndex((page) => page?.pageNumber === parseInt(newPage))
+    if (currentPageIndex > -1) {
+      const after = pages[currentPageIndex]?.after
+      setQueryOptions("ComponentInstancesOfService", {
+        ...queryOptions,
+        after: `${after}`,
+      })
+    }
+  }
+
+  return (
+    <>
+      <ContentHeading className="mt-8 mb-2" heading="Component Instances" />
+      <DataGrid columns={4}>
+        <DataGridRow>
+          <DataGridHeadCell>Component</DataGridHeadCell>
+          <DataGridHeadCell>Version</DataGridHeadCell>
+          <DataGridHeadCell>Total Number of Issues</DataGridHeadCell>
+          <DataGridHeadCell>Highest Severity</DataGridHeadCell>
+        </DataGridRow>
+        {isLoading ? (
+          <DataGridRow>
+            <DataGridCell colSpan={4}>
+              <Container py>
+                <LoadElement />
+              </Container>
+            </DataGridCell>
+          </DataGridRow>
+        ) : (
+          items.map((componentInstance, i) => (
+            <DataGridRow key={i}>
+              <DataGridCell>{componentInstance?.node?.ccrn}</DataGridCell>
+              <DataGridCell className="break-all overflow-hidden">
+                {componentInstance?.node?.componentVersion?.version}
+              </DataGridCell>
+              <DataGridCell>{componentInstance?.node?.issueMatches?.totalCount}</DataGridCell>
+              <DataGridCell>
+                {severityString(highestSeverity(componentInstance?.node?.issueMatches?.edges))}
+              </DataGridCell>
+            </DataGridRow>
+          ))
+        )}
+      </DataGrid>
+
+      <Container py px={false}>
+        <Stack className="flex justify-end">
+          <Pagination
+            currentPage={currentPage}
+            isFirstPage={currentPage === 1}
+            isLastPage={currentPage === totalPages}
+            onPressNext={() => onPaginationChanged(currentPage + 1)}
+            onPressPrevious={() => onPaginationChanged(currentPage - 1)}
+            onKeyPress={(oKey) => {
+              if (oKey.code === "Enter") {
```

Comment:
Maybe extract to a naming function here what do you think?

Reply:
Do you mean to define different functions for each of above functions? I do not see a big advantage of that to be honest. So I leave it as it is, if you don't mind.